[feature-hal-itm] Instrumented Trace Macrocell HAL API for SWO debug output

marcuschangarm opened the pull request:
HAL API for initializing ITM and setting up SWO output. 

Uses FileHandle redirection for debug output.

Note: this should be merged to a new feature branch and not to master.

Depends on: https://github.com/ARMmbed/mbed-os/pull/5571

marcuschangarm commented:
@yogpan01 @TeroJaasko  

This is what you really need for the miniclient.

marcuschangarm commented:
@0xc0170 @bulislaw 

As we discussed. Can you create a feature branch please?

kjbracey-arm commented on platform/mbed_itm.h:
```diff
@@ -0,0 +1,58 @@
+
+#include "hal/itm_api.h"
+
+class SingleWireOutput : public FileHandle {
```
To fit the pattern of the rest of the system, this would be drivers/SingleWireOutput.h - you've put the ITM bits in the HAL, and this is the corresponding C++ wrapper. So it goes with SPI and friends.

marcuschangarm replied:
I was hoping to eventually put it into mbed_retarget.cpp like you did with the serial class: https://github.com/kjbracey-arm/mbed-os/blob/d7e42d1f69e5a8f06441742d910d795c8ae22c85/platform/mbed_retarget.cpp#L128-L180

so that we can enable SWO from the config system. 

My intention with this PR is to start some discussion about what would be right approach. @0xc0170 @bulislaw any comments?

kjbracey-arm replied:
I'm slightly wary of setting up a sort of "central registry" pattern where every potential device has to be plopped in there.

In #5558 there's a similar concept - the JSON choice of "default onboard network stack". You set `nsapi.default-stack` to either "LWIP" or "NANOSTACK". There's no central logic for that, but each of the two stacks has a code snippet that looks like

    #define NANOSTACK 0x82917429
    #if MBED_CONF_NSAPI_DEFAULT_STACK == NANOSTACK
    OnboardNetworkStack &OnboardNetworkStack::get_default_instance() {
        return Nanostack::get_instance();
    }
   #endif

So only one of them will define that function.

Maybe the same concept would work here. JSON option in central place, but any target or library one can add a new selectable default without touching the core code - just invent a name.

kjbracey-arm replied:
Anyway, regardless of the "default console" selection, this needs to be exposed as a separate standalone device manually accessible as `SingleWireOutput`.

See my 18 Dec comments on #5356, where they're concentrating on alternative console, but I'm urging them to make sure that the device is always manually useable by the app without being forced to make it stdin/stdout.

The built-in `FileHandle`s in my mbed_retarget.cpp are purely to preserve legacy behaviour. If anything, they will be moved out (@geky has suggested making `Sink` public, and discussions with @sg- have led to concensus that `RawSerial` should be a `FileHandle`, thus it can do the job of `DirectSerial`).

marcuschangarm replied:
No problem, I'll move/rename it to mbed-os/drivers. 

I'm fine with either invocation as well, but I think we should have it documented exactly how we expect people to do it!

kjbracey-arm commented on platform/mbed_itm.h:
```diff
@@ -0,0 +1,58 @@
+
+    virtual ssize_t read(void *buffer, size_t size) {
+        unsigned char *buf = static_cast<unsigned char *>(buffer);
+        buf[0] = 0;
```
I think it's better to just return 0 (indicating "end of file") or an error code (not sure which - -EBADF = "not valid file descriptor open for reading"?; ENXIO = "... request was outside the capabilities of the device"?). Not suire which of those two choices I prefer.

My dummy device implementation inside the retarget code is only reading endless NUL bytes to preserve the legacy behaviour of stdin with DEVICE_SERIAL unset. I don't think that's actually useful, or what you should do for something new.

kjbracey-arm replied:
Did a bit of background reading to compare other systems. This is an interesting general question.

`/dev/null` returns EOF if opened for input. But that's more likely to be used manually as a redirect for processes that expect a redirect from file, and hence understand/expect EOF on stdin.

Many real devices probably would do `ENXIO` on the attempt to open for read. I guess that means attempts to use a read-only device as process input would fail the read open and then you'd get `EBADF` later.

Linux gives `EBADF` if you read from a file that's open for output only.

When you `nohup` and detach a process in Linux it does seem you get `EBADF` on input (and you can change that to EOF if you redirect from /dev/null). POSIX says behaviour of nohup input is up to the implementation - it might leave input from original terminal, or get it from a file.

On balance, I guess I'll say `EBADF` is what to do on read. This is a "standalone" FileHandle where it's automatically opened on construction, so we should assume that the implicit open is write-only, thus give `EBADF` on read attempts.

If there was a device filing system, so we were able to participate in the open call, we'd return ENXIO on the attempt to open for read.

marcuschangarm replied:
Thank you! I'll change it to `EBADF`.

geky replied:
Lesson learned, looks like I have some filesystems to update...

kjbracey-arm replied:
Ah, normal filesystems might be a bit different. Eg EACCES for access violation rather than ENXIO. I was only really thinking about devices, and particularly considering this odd case of "opened by construction" objects.

geky replied:
Well, in the open group docs for write:
http://pubs.opengroup.org/onlinepubs/9699919799/functions/write.html

> [EBADF]
> The fildes argument is not a valid file descriptor open for writing.

Pretty straightforward for POSIX docs actually. 😆 

A quick test program confirms this is what Linux returns.

Fortunately reading/writing to a file after opening it in a specific mode is pretty uncommon outside of user errors. I'd be worried if someone was relying on a specific error code in that situation.

kjbracey-arm commented on platform/mbed_itm.h:
```diff
@@ -0,0 +1,58 @@
+    virtual ssize_t write(const void *buffer, size_t size) {
+        const unsigned char *buf = static_cast<const unsigned char *>(buffer);
+        for (size_t i = 0; i < size; i++) {
+            ITM_SendChar(buf[i]);
```
This isn't in itm_api.h? Skipping a layer?

marcuschangarm replied:
ITM_SendChar is defined in:

```
cmsis/TARGET_CORTEX_M/core_cm3.h
cmsis/TARGET_CORTEX_M/core_cm4.h
etc.
```

I'm not sure how to explicitly include the right header file based on core.

marcuschangarm replied:
ah, `#include "cmsis.h"` should do the trick.

0xc0170 replied:
yes, cmsis should be used as a generic cmsis header file

TeroJaasko commented:
Works on my NRF52_DK. Redirecting the console is now really easy:
``` 
FileHandle *mbed::mbed_override_console(int fd) {
    static SingleWireOutput swo_serial;
    return &swo_serial;
}

int main() {
    while (true) {
        wait(0.5);
        printf("hello world\n");
    }
}
```

pan- commented on platform/mbed_itm.h:
```diff
@@ -0,0 +1,58 @@
+
+#if defined(DEVICE_ITM)
+
+#include "hal/itm_api.h"
```
I think you'll need to include `platform/FileHandle.h`.

marcuschangarm replied:
Thank you! I'll add the missing include.

yogpan01 commented:
@marcuschangarm Thanks for this ! @0xc0170 When can this be merged into MASTER. We are currently blocked on using nRF52 for our development because of this. We need this ASAP !
@JanneKiiskila @sg- FYI

0xc0170 commented:
> @0xc0170 When can this be merged into MASTER

First, Filehandle PR needs to be integrated first (there are still some reviews + tests outstanding). As soon as that one is in, this should be updated to reflect the latest changes.

marcuschangarm commented:
Rebased with master to get https://github.com/ARMmbed/mbed-os/pull/5571

marcuschangarm commented:
@0xc0170 @yogpan01 

>>@0xc0170 When can this be merged into MASTER

>First, Filehandle PR needs to be integrated first (there are still some reviews + tests outstanding). As soon as that one is in, this should be updated to reflect the latest changes.

FileHandle PR has been merged and this PR has been rebased. Now what? 😄

kjbracey-arm commented:
/morph build

kjbracey-arm commented:
@SenRamakri, @c1728p9, @bulslaw - dependency is now merged, so this could precede. Please review.

I'm happy with the `FileHandle` end of this. Are people happy with the HAL API?

bulislaw commented on hal/itm_api.h:
```diff
@@ -0,0 +1,34 @@
+extern "C" {
+#endif
+
+void itm_init(void);
```
I know it's a dummy function but could we get some description? What should it initialise (channel 0?). Here's some example https://github.com/ARMmbed/mbed-os/blob/feature-hal-spec-rtc/hal/rtc_api.h also can you add it to doxy please.
@c1728p9 any comments?

bulislaw commented on drivers/SerialWireOutput.h:
```diff
@@ -0,0 +1,58 @@
+#include "hal/itm_api.h"
+#include "platform/FileHandle.h"
+
+class SerialWireOutput : public FileHandle {
```
Not a single comment :P @SenRamakri please review

marcuschangarm replied:
Seemed a bit premature until we've all agreed on the functionality! 😄

SenRamakri replied:
@marcuschangarm - Since this is coming into master its better to have some comments/documentation on what this class is?

SenRamakri commented on targets/TARGET_Silicon_Labs/TARGET_EFM32/itm_api.c:
```diff
@@ -0,0 +1,114 @@
+    CMU->OSCENCMD = CMU_OSCENCMD_AUXHFRCOEN;
+  
+    // Wait until clock is ready
+    while (!(CMU->STATUS & CMU_STATUS_AUXHFRCORDY));
```
Potentially infinite loop, can we please break this loop and return with an error after waiting enough? I don't how much is "enough", you can define that.

marcuschangarm replied:
All Silicon Labs clock initialization code use this kind of infinite loop. Changing this single loop with an arbitrary loop counter would break the pattern.

SenRamakri commented on targets/TARGET_Silicon_Labs/TARGET_EFM32/itm_api.c:
```diff
@@ -0,0 +1,114 @@
+    uint32_t div  = freq / 875000;
+    TPI->ACPR = div - 1;
+
+    ITM->LAR  = 0xC5ACCE55;
```
It would be better if we can #define these magic values.

SenRamakri commented on targets/TARGET_Silicon_Labs/TARGET_EFM32/itm_api.c:
```diff
@@ -0,0 +1,114 @@
+    ITM->TCR  = 0x0;
+    TPI->SPPR = 2;
+    ITM->TPR  = 0x0;
+    DWT->CTRL = 0x400003FF;
```
Magic value 0x400003FF and others below. Can we please define them with corresponding fields? Also it looks like, we are writing to upper nibble which is read-only I guess. We are also enabling CYCCNTENA, POSTINIT etc, are these required? May be we should review this value as to what all we need to enable.

marcuschangarm commented:
@bulislaw 

I've added comments and wrote a short porting guide: https://github.com/ARMmbed/Handbook/pull/413

@SenRamakri 

I've moved the generic ITM register settings out into a separate `mbed_itm_api.c` file so it can be shared. Each target `itm_api.c` will only be responsible for setting up it's own clock and SWO pin. I've also replaced the magic numbers with named bit positions.

marcuschangarm commented:
/morph build

marcuschangarm commented:
/morph build

marcuschangarm commented:
/morph build
